pennylane/labs/resource_estimation/resource_operator.py

```diff
@@ -0,0 +1,530 @@
+
+    @property
+    @abstractmethod
+    def resource_params(self) -> dict:
+        """A dictionary containing the minimal information needed to compute a resource estimate
+        of the operator's decomposition. The keys of this dictionary should match the
+        ``resource_keys`` attribute of the operator class.
+        """
+
+    @classmethod
+    @abstractmethod
+    def resource_rep(cls, *args, **kwargs):
+        r"""Returns a compressed representation containing only the parameters of
+        the Operator that are needed to estimate the resources."""
+
+    def resource_rep_from_op(self):
+        r"""Returns a compressed representation directly from the operator"""
+        return self.__class__.resource_rep(**self.resource_params)
+
+    @classmethod
+    @abstractmethod
+    def default_resource_decomp(cls, *args, **kwargs) -> List:
+        r"""Returns a list of actions that define the resources of the operator."""
+
+    @classmethod
+    def resource_decomp(cls, *args, **kwargs) -> List:
+        r"""Returns a list of actions that define the resources of the operator."""
+        return cls.default_resource_decomp(*args, **kwargs)
+
+    @classmethod
+    def default_adjoint_resource_decomp(cls, *args, **kwargs) -> List:
+        r"""Returns a list representing the resources for the adjoint of the operator."""
+        raise ResourcesNotDefined
+
+    @classmethod
+    def adjoint_resource_decomp(cls, *args, **kwargs) -> List:
+        r"""Returns a list of actions that define the resources of the operator."""
+        return cls.default_adjoint_resource_decomp(*args, **kwargs)
+
+    @classmethod
+    def default_controlled_resource_decomp(
+        cls, ctrl_num_ctrl_wires: int, ctrl_num_ctrl_values: int, *args, **kwargs
+    ) -> List:
+        r"""Returns a list representing the resources for a controlled version of the operator.
+
+        Args:
+            ctrl_num_ctrl_wires (int): the number of qubits the
+                operation is controlled on
+            ctrl_num_ctrl_values (int): the number of control qubits, that are
+                controlled when in the :math:`|0\rangle` state
+        """
+        raise ResourcesNotDefined
+
+    @classmethod
+    def controlled_resource_decomp(
+        cls, ctrl_num_ctrl_wires: int, ctrl_num_ctrl_values: int, *args, **kwargs
+    ) -> List:
+        r"""Returns a list representing the resources for a controlled version of the operator.
+
+        Args:
+            ctrl_num_ctrl_wires (int): the number of qubits the
+                operation is controlled on
+            ctrl_num_ctrl_values (int): the number of control qubits, that are
+                controlled when in the :math:`|0\rangle` state
+        """
+        return cls.default_controlled_resource_decomp(
+            ctrl_num_ctrl_wires, ctrl_num_ctrl_values, *args, **kwargs
+        )
+
+    @classmethod
+    def default_pow_resource_decomp(cls, pow_z: int, *args, **kwargs) -> List:
```

Comment:
Is pow_z use anywhere?

Reply:
It will be used when we add pow decompositions to our resource operators. Otherwise I haven't seen it used in core PL.